vpr/src/pack/pb_type_graph.cpp

```diff
@@ -334,6 +350,179 @@ static void alloc_and_load_pb_graph(t_pb_graph_node* pb_graph_node,
+                    auto curr_pb_graph_pin = &pins[port_idx][pin_idx];
+                    pb_pin_idx_map.insert(std::make_pair(curr_pb_graph_pin->pin_count_in_cluster, curr_pb_graph_pin));
+                }
+            }
+        }
+        /* logical indices for the next block should be shifted by the number of pins seen so far. As a result, the given logical index will be unique in the logical_block level*/
+        offset += curr_pb_type->num_pins;
+
+        if(curr_pb_graph_node->is_root()){
+            VTR_ASSERT(offset == curr_pb_graph_node->pb_type->num_pins);
+        }
+
+        /* Push sub-blocks to the queue */
+        for(int mode_idx = 0; mode_idx < curr_pb_type->num_modes; mode_idx++) {
+            auto curr_mode = &curr_pb_type->modes[mode_idx];
+            for (int pb_type_idx = 0; pb_type_idx < curr_mode->num_pb_type_children; pb_type_idx++) {
+                int num_pb = curr_mode->pb_type_children[pb_type_idx].num_pb;
+                for (int pb_idx = 0; pb_idx < num_pb; pb_idx++) {
+                    auto child_pb_graph_node = &(curr_pb_graph_node->child_pb_graph_nodes[mode_idx][pb_type_idx][pb_idx]);
+                    pb_graph_node_to_set_pin_index_q.push(child_pb_graph_node);
+                }
+            }
+        }
+    }
+
+}
+
+static std::vector<const t_pb_graph_node*> get_all_logical_block_pb_graph_nodes(const t_pb_graph_node* pb_graph_node) {
+    std::vector<const t_pb_graph_node*> pb_graph_nodes;
+    pb_graph_nodes.push_back(pb_graph_node);
+    if(pb_graph_node->is_primitive()) {
+        return pb_graph_nodes;
+    }
+
+    auto pb_type = pb_graph_node->pb_type;
+    for(int mode_idx = 0; mode_idx < pb_graph_node->pb_type->num_modes; mode_idx++) {
+        for(int pb_type_idx = 0; pb_type_idx < (pb_type->modes[mode_idx]).num_pb_type_children; pb_type_idx++) {
+            int num_pb = pb_type->modes[mode_idx].pb_type_children[pb_type_idx].num_pb;
+            for(int pb_idx = 0; pb_idx < num_pb; pb_idx++) {
+                const t_pb_graph_node* child_pb_graph_node = &(pb_graph_node->child_pb_graph_nodes[mode_idx][pb_type_idx][pb_idx]);
+                auto sub_pb_graph_nodes = get_all_logical_block_pb_graph_nodes(child_pb_graph_node);
+                pb_graph_nodes.insert(std::end(pb_graph_nodes), std::begin(sub_pb_graph_nodes), std::end(sub_pb_graph_nodes));
+
+            }
+        }
+    }
+    return pb_graph_nodes;
+}
+
+static void add_logical_classes(t_logical_block_type* logical_block) {
+
+    logical_block->pb_pin_class_map = std::unordered_map<const t_pb_graph_pin*, int>();
+
+    auto pb_graph_nodes = get_all_logical_block_pb_graph_nodes(logical_block->pb_graph_head);
+    for(auto pb_graph_node : pb_graph_nodes) {
+        for(int port_type = 0; port_type < 3; port_type++) {
+            int num_ports;
+            int* num_pins;
+            t_pb_graph_pin** pb_graph_pins;
+            if(port_type == 0) {
```

Comment:
use IN_PORT , OUT_PORT instead of 0,1,2

Reply:
There is no CLOCK_PORT Type. There are also other functions that follow the same convention. If you think it is useful, I can make a separate struct to specify the logical port type.